File: src/images_split.py
import numpy as np

def images_split(Storage, type_pass, multigrid, borders):
    """
    images_split Calculate the coordinates of the interrogation windows on the image
    Performs the calculation of the coordinates of the interrogation windows without considering the vector field.
    """
    
    if type_pass == 'first' or multigrid:  # calculate the coordinates of the interrogation windows
        image_size = Storage.image_1.shape
        
        image_size = np.array(image_size)
        window_size = np.array(Storage.window_size)
        # Number of interrogation windows horizontally and vertically
        field_shape = np.floor((image_size - window_size) / (window_size - Storage.overlap)) + 1
        
        # Size of the area covered by the interrogation windows
        w = field_shape[1] * (Storage.window_size[1] - Storage.overlap[1]) + Storage.overlap[1]
        h = field_shape[0] * (Storage.window_size[0] - Storage.overlap[0]) + Storage.overlap[0]
        
        # Residual boundaries not covered by the interrogation windows
        remainder = [image_size[0] - h, image_size[1] - w]
        
        # Residual boundaries in the upper left part of the image
        x_left_remainder = np.floor(remainder[1] / 2) + 1
        y_top_remainder = np.floor(remainder[0] / 2) + 1
        
        # Size of the centered area covered by the interrogation windows
        search_area_size = [x_left_remainder, y_top_remainder, w, h]
        
        # Calculate the centers of the interrogation windows
        if (Storage.window_size[1] - Storage.overlap[1]) > 1:  # for overlap greater than one pixel
            x = np.arange(search_area_size[0] + np.ceil(Storage.window_size[1] / 2) - 1, 
                          search_area_size[0] + search_area_size[2] - np.ceil(Storage.window_size[1] / 2) + 1, 
                          Storage.window_size[1] - Storage.overlap[1])
            y = np.arange(search_area_size[1] + np.ceil(Storage.window_size[0] / 2) - 1, 
                          search_area_size[1] + search_area_size[3] - np.ceil(Storage.window_size[0] / 2) + 1, 
                          Storage.window_size[0] - Storage.overlap[0])
        else:  # for overlap equal to one pixel
            x = np.arange(search_area_size[0] + np.ceil(Storage.window_size[1] / 2) - 1, 
                          search_area_size[0] + search_area_size[2] - np.ceil(Storage.window_size[1] / 2), 
                          Storage.window_size[1] - Storage.overlap[1])
            y = np.arange(search_area_size[1] + np.ceil(Storage.window_size[0] / 2) - 1, 
                          search_area_size[1] + search_area_size[3] - np.ceil(Storage.window_size[0] / 2), 
                          Storage.window_size[0] - Storage.overlap[0])
        
        # Add interrogation windows at the borders of the image for full coverage
        if borders:
            if remainder[0] != 0:
                if remainder[0] == 1:  # if the residual boundary is one pixel
                    y = np.append(y, y[-1] + np.floor(remainder[0] / 2) + np.mod(remainder[0], 2))
                else:
                    y = np.append(np.insert(y, 0, y[0] - np.floor(remainder[0] / 2)), y[-1] + np.floor(remainder[0] / 2) + np.mod(remainder[0], 2))
            if remainder[1] != 0:
                if remainder[1] == 1:  # if the residual boundary is one pixel
                    x = np.append(x, x[-1] + np.floor(remainder[1] / 2) + np.mod(remainder[1], 2))
                else:
                    x = np.append(np.insert(x, 0, x[0] - np.floor(remainder[1] / 2)), x[-1] + np.floor(remainder[1] / 2) + np.mod(remainder[1], 2))
        
        X0, Y0 = np.meshgrid(x, y)
        return X0, Y0

# Example usage
# storage = Storage()
# storage.image_1 = np.random.rand(100, 100)
# storage.window_size = np.array([32, 32])
# storage.overlap = np.array([16, 16])
# X0, Y0 = images_split(storage, 'first', True, True)

File: matlab/images_split.m
function [X0,Y0] = images_split(Storage,type_pass,multigrid,borders)
%images_split Расчет координат окон опроса на изображении
%   Выполняет расчет координат окон опроса без учета векторного поля

if strcmp(type_pass,'first') || (multigrid) % рассчитываем координаты положения окон опроса
    image_size = size(Storage.image_1);
    
    % Количество окон опроса по горизонтали и вертикали
    field_shape = floor((image_size - Storage.window_size)./(Storage.window_size - Storage.overlap)) + 1;
    
    % Размер области покрытия изображения окнами опроса
    w = field_shape(2)*(Storage.window_size(2) - Storage.overlap(2))+Storage.overlap(2);
    h = field_shape(1)*(Storage.window_size(1) - Storage.overlap(1))+Storage.overlap(1);
    
    % Остаточные границы не покрытых окнами опроса
    remainder = [image_size(1) - h, image_size(2) - w];
    
    % Остаточные границы в левой верхней части изображения
    x_left_remainder = floor(remainder(2)/2)+1;
    y_top_remainder = floor(remainder(1)/2)+1;
    
    % Размер центрированной области покрытия окнами опроса изображения
    search_area_size = [x_left_remainder, y_top_remainder, w, h];
    
    % Вычисление центров окон опроса
    if (Storage.window_size(2) - Storage.overlap(2)) > 1 % для наложения больше одного пикселя
        x = (search_area_size(1)+ceil(Storage.window_size(2)/2) - 1):(Storage.window_size(2) - Storage.overlap(2)):(search_area_size(1)+search_area_size(3)-ceil(Storage.window_size(2)/2));
        y = (search_area_size(2)+ceil(Storage.window_size(1)/2) - 1):(Storage.window_size(1) - Storage.overlap(1)):(search_area_size(2)+search_area_size(4)-ceil(Storage.window_size(1)/2));
    else                                                 % для наложения равного одному пикселю
        x = (search_area_size(1)+ceil(Storage.window_size(2)/2) - 1):(Storage.window_size(2) - Storage.overlap(2)):(search_area_size(1)+search_area_size(3)-ceil(Storage.window_size(2)/2) - 1);
        y = (search_area_size(2)+ceil(Storage.window_size(1)/2) - 1):(Storage.window_size(1) - Storage.overlap(1)):(search_area_size(2)+search_area_size(4)-ceil(Storage.window_size(1)/2) - 1);
    end
    
    % Добавление окон опроса на границах изображения для полного покрытия изображения
    if borders
        if remainder(1) ~= 0
            if remainder(1) == 1 % если остаточная граница равна одному пикселю
                y = [y(:);y(end) + floor(remainder(1)/2) + mod(remainder(1),2)];
            else
                y = [y(1) - floor(remainder(1)/2);y(:);y(end) + floor(remainder(1)/2) + mod(remainder(1),2)];
            end
        end
        if remainder(2) ~= 0
            if remainder(2) == 1
                x = [x(:);x(end) + floor(remainder(2)/2) + mod(remainder(2),2)];
            else
                x = [x(1) - floor(remainder(2)/2);x(:);x(end) + floor(remainder(2)/2) + mod(remainder(2),2)];
            end
        end
    end
    
    % Определение матриц координат центров окон опроса
    [X,Y] = meshgrid(x,y);
    
    % Запись центров окон опроса
    Storage.centers_map = zeros([size(X),2]);
    Storage.centers_map(:,:,1) = X;
    Storage.centers_map(:,:,2) = Y;
    
    % Запись начальных координат окн опроса, начало которых находится в левом
    % верхнем углу окна
    X0 = floor(X)-ceil(Storage.window_size(2)/2) + 1;
    Y0 = floor(Y)-ceil(Storage.window_size(1)/2) + 1;
else % если не требуется расчет новых координат окон опроса
    X0 = floor(Storage.centers_map(:,:,1))-ceil(Storage.window_size(2)/2) + 1;
    Y0 = floor(Storage.centers_map(:,:,2))-ceil(Storage.window_size(1)/2) + 1;
end

end